### PR TITLE
Make stash a builtin

### PR DESCRIPTION
**Pull request by rimrul**
These are the first two Patches in an upcomming series of Patches to
convert git stash into a builtin command. This is mainly based on the
general performance of scripts running inside an interpreter (bash in
this case) running inside a POSIX emulation layer on Windows. Shell,
perl and python scripts are generaly a lot faster on Unix-like systems
compared to Windows systems. That does not mean that Unix-like systems
won't benefit from more native Git commands, but the effect is bigger
on Git for Windows. 

These two patches port over the core non-patch part of git-stash into
native code using a separate helper command. This helper command is
intended as a temporary meassure and as such it's subject to change.
For this reason, I did not implement new regression tests, documentation
or localizations for this command. 

I meassured the changes in excecution time for the stash related
regression tests on the same hardware running Windows 8.1 and
Kubuntu 15.10. Each result is the difference between the average of five
meassurements (six on Linux, because I lost count on the first run of
meassurements) each before and after these changes. I meassured the
following changes:

Windows:

t3903
real  -5,10%
user -0,94%
sys  +0,16% (10ms)

t3904
real  -0,30%
user -2,98% (20ms)
sys  +5,03%

t3905
real  -4,03%
user -8,13%
sys +17,42%

t3906
real  -2,57%
user +1,94%
sys   +1,59%

Linux:

t3903
real    +0,63%
user +10,87% (3ms)
sys     +4,29% (4ms)

t3904
real    -7,29%
user -30,61%
sys   +5,77% (4ms)

t3905
real  -7,29%
user -33,33% (2ms)
sys   +20% (2ms)

t3906
real  -0,88%
user -1,08% (1ms)
sys   -2,22%

I added the asolute times where I think the difference is below the
meassurement precission (4ms on Linux) and on the two lowest absolute
differences on windows. A full log of all meassurement runs is available
at https://gist.github.com/rimrul/82adf3b368ed633263d2. Please note that
according to Johannes Schindelin, maintainer of Git for Windows, the
meassuring of sys time on Windows is unreliable. With that in mind,
in summary this is a slight increase in performance on Linux, and a more
noticeable increase on Windows.

### CODE REVIEW

**Comment by submitgit** [comment]
@rimrul sent these 2 commits (https://github.com/git/git/compare/d10e2cb9d0299a26f43d57dd5bdcf2b3f86a30b3...0027f17f392adf8b014113e6718792788b4c06fa) as a set of patches to the mailing list with [_submitGit_](https://github.com/rtyley/submitgit) - here on [Gmane](http://mid.gmane.org/0000015289cc78a7-0cb0335d-d8a2-4e8c-a672-d400b86de9fa-000000@eu-west-1.amazonses.com), [MARC](http://marc.info/?i=0000015289cc78a7-0cb0335d-d8a2-4e8c-a672-d400b86de9fa-000000@eu-west-1.amazonses.com) [](0000015289cc78a7-0cb0335d-d8a2-4e8c-a672-d400b86de9fa-000000@eu-west-1.amazonses.com)

**Comment by submitgit** [comment]
@rimrul sent these 2 commits (https://github.com/git/git/compare/d10e2cb9d0299a26f43d57dd5bdcf2b3f86a30b3...0027f17f392adf8b014113e6718792788b4c06fa) as a set of patches to the mailing list with [_submitGit_](https://github.com/rtyley/submitgit) - here on [Gmane](http://mid.gmane.org/0000015289e00b84-827747ee-d170-42bf-ab1e-6e2893bdc702-000000@eu-west-1.amazonses.com), [MARC](http://marc.info/?i=0000015289e00b84-827747ee-d170-42bf-ab1e-6e2893bdc702-000000@eu-west-1.amazonses.com) [](0000015289e00b84-827747ee-d170-42bf-ab1e-6e2893bdc702-000000@eu-west-1.amazonses.com)

**Comment by submitgit** [comment]
@rimrul sent these 2 commits (https://github.com/git/git/compare/d10e2cb9d0299a26f43d57dd5bdcf2b3f86a30b3...0027f17f392adf8b014113e6718792788b4c06fa) as a set of patches to the mailing list with [_submitGit_](https://github.com/rtyley/submitgit) - here on [Gmane](http://mid.gmane.org/0000015289f33df4-d0095101-cfc0-4c41-b1e7-6137105b93fb-000000@eu-west-1.amazonses.com), [MARC](http://marc.info/?i=0000015289f33df4-d0095101-cfc0-4c41-b1e7-6137105b93fb-000000@eu-west-1.amazonses.com) [](0000015289f33df4-d0095101-cfc0-4c41-b1e7-6137105b93fb-000000@eu-west-1.amazonses.com)

**Comment by rimrul** [comment]
The pull request has reached the [mailing list](http://thread.gmane.org/gmane.comp.version-control.git/285022) after some trouble related to rtyley/submitgit#23.

**Comment by dscho** [comment]
I guess this can be closed, now that a built-in `stash` is on its way to v2.22.0?

**Comment by rimrul** [comment]
Yes, I agree

**Comment by rimrul** [comment]
Oh, yes, I should've closed this back in April 2018 along with git-for-windows/git#508

**Comment by dscho** [comment]
:-)
